src/Workspaces/Core/Portable/Workspace/Solution/Project.cs

```diff
@@ -254,6 +256,19 @@ public bool ContainsAnalyzerConfigDocument(DocumentId documentId)
             return ImmutableHashMapExtensions.GetOrAdd(ref _idToAnalyzerConfigDocumentMap, documentId, s_createAnalyzerConfigDocumentFunction, this);
         }
 
+        public async ValueTask<IEnumerable<SourceGeneratedDocument>> GetSourceGeneratedDocumentsAsync(CancellationToken cancellationToken = default)
+        {
+            var generatedDocumentStates = await _solution.State.GetSourceGeneratedDocumentStatesAsync(this.State, cancellationToken).ConfigureAwait(false);
+            using var _ = ArrayBuilder<SourceGeneratedDocument>.GetInstance(generatedDocumentStates.Length, out var builder);
+
+            foreach (var generatedDocumentState in generatedDocumentStates)
+            {
+                builder.Add(ImmutableHashMapExtensions.GetOrAdd(ref _idToSourceGeneratedDocumentMap, generatedDocumentState.Id, s_createSourceGeneratedDocumentFunction, (generatedDocumentState, this)));
+            }
+
+            return builder.ToImmutable();
```

Comment:
looking at this... this all seems threadsafe...  i assume the ImmutableHashMapExtensions is appropriately threadsafe though?  i.e. say there are two threads that are calling into this method, and now both are in the foreach.  will the GetOrAdd work properly for two concurrent mutations?

Reply:
Yes, it's a crazy method that does all sorts of fancy interlocked reading. It's the same method we use for the other GetDocument and friends.